src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/BStrStringMarshaller.cs

```diff
@@ -0,0 +1,120 @@
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices.Marshalling
+{
+    /// <summary>
+    /// Marshaller for BSTR strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x200,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct BStrStringMarshaller
+    {
+        private byte* _ptrToFirstChar;
+        private bool _allocated;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BStrStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public BStrStringMarshaller(string? str)
+            : this(str, default)
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BStrStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// The <paramref name="buffer"/> must not be movable - that is, it should not be
+        /// on the managed heap or it should be pinned.
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public BStrStringMarshaller(string? str, Span<byte> buffer)
+        {
+            _allocated = false;
+
+            if (str is null)
+            {
+                _ptrToFirstChar = null;
+                return;
+            }
+
+            int lengthInBytes = checked(sizeof(char) * str.Length);
+
+            // A caller provided buffer must be at least (lengthInBytes + 6) bytes
+            // in order to be constructed manually. The 6 extra bytes are 4 for byte length and 2 for wide null.
+            if ((lengthInBytes + 6) > buffer.Length)
+            {
+                // Use precise byte count when the provided stack-allocated buffer is not sufficient
+                _ptrToFirstChar = (byte*)Marshal.AllocBSTRByteLen((uint)lengthInBytes);
+                _allocated = true;
+            }
+            else
+            {
+                // set length and update buffer target
+                byte* pBuffer = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
+                *((uint*)pBuffer) = (uint)lengthInBytes;
+                _ptrToFirstChar = pBuffer + sizeof(uint);
+            }
+
+            // Confirm the size is properly set for the allocated BSTR.
+            Debug.Assert(lengthInBytes == Marshal.SysStringByteLen((IntPtr)_ptrToFirstChar));
+
+            // Copy characters from the managed string
+            Buffer.Memmove(ref *(char*)_ptrToFirstChar, ref str.GetRawStringData(), (nuint)str.Length + 1);
```

Comment:
```suggestion
            str.CopyTo(new Span<char>(_ptrToFirstChar, str.Length));
            _ptrToFirstChar[str.Length] = '\0'; // null-terminate
```
For symmetry with `Utf16StringMarshaller`. Also, it looks better to use public methods rather than the internal helper methods.